beacon_node/http_api/src/attestation_performance.rs

```diff
@@ -0,0 +1,210 @@
+    ParticipationCache(ParticipationCacheError),
+    UnableToFindValidator(usize),
+}
+
+impl From<BlockReplayError> for AttestationPerformanceError {
+    fn from(e: BlockReplayError) -> Self {
+        Self::BlockReplay(e)
+    }
+}
+
+impl From<BeaconStateError> for AttestationPerformanceError {
+    fn from(e: BeaconStateError) -> Self {
+        Self::BeaconState(e)
+    }
+}
+
+impl From<ParticipationCacheError> for AttestationPerformanceError {
+    fn from(e: ParticipationCacheError) -> Self {
+        Self::ParticipationCache(e)
+    }
+}
+
+pub fn get_attestation_performance<T: BeaconChainTypes>(
+    target: String,
+    query: AttestationPerformanceQuery,
+    chain: Arc<BeaconChain<T>>,
+) -> Result<Vec<AttestationPerformance>, warp::Rejection> {
+    let spec = &chain.spec;
+    // We increment by 2 here so that when we build the state from the `prior_slot` it is
+    // still 1 epoch ahead of the first epoch we want to analyse.
+    // This ensures the `.is_previous_epoch_X` functions on `EpochProcessingSummary` return results
+    // for the correct epoch.
+    let start_epoch = query.start_epoch + 2;
+    let start_slot = start_epoch.start_slot(T::EthSpec::slots_per_epoch());
+    let prior_slot = start_slot - 1;
+
+    let end_epoch = query.end_epoch + 2;
+    let end_slot = end_epoch.end_slot(T::EthSpec::slots_per_epoch());
+
+    // Check query is valid
+    if start_epoch > end_epoch {
+        // || start_epoch == 0 {
+        return Err(custom_bad_request(format!(
+            "invalid start and end epochs: {}, {}",
+            query.start_epoch, query.end_epoch
+        )));
+    }
+
+    // The response size can grow exceptionally large therefore we should check that the
+    // query is within permitted bounds to prevent potential OOM errors.
+    if (end_epoch - start_epoch).as_usize() > MAX_REQUEST_RANGE_EPOCHS {
+        return Err(custom_bad_request(format!(
+            "end_epoch must not exceed start_epoch by more than 100 epochs. start: {}, end: {}",
+            query.start_epoch, query.end_epoch
+        )));
+    }
+
+    let head_state = chain.head_beacon_state().map_err(beacon_chain_error)?;
+
+    let validator_indices = (0..=head_state.validators().len() as u64).collect();
+
+    // Either use the global validator set, or the specified index.
+    let index_range = if target.to_lowercase() == "global" {
+        validator_indices
+    } else {
+        vec![target.parse::<u64>().map_err(|_| {
+            custom_bad_request(format!(
+                "invalid validator index: {:?}",
+                target.to_lowercase()
+            ))
+        })?]
+    };
+
+    // Load block roots.
+    let mut block_roots: Vec<Hash256> = chain
+        .forwards_iter_block_roots_until(start_slot, end_slot)
+        .map_err(beacon_chain_error)?
+        .collect::<Result<Vec<(Hash256, Slot)>, _>>()
+        .map_err(beacon_chain_error)?
+        .iter()
+        .map(|(root, _)| *root)
+        .collect();
```

Comment:
If we do the `|(root, _)| root` mapping before collecting the first time we can avoid some allocations here. It'll end up being a double map like:

```rust
.map(|res| res.map(|(root, _)| root))
```